{
 "metadata": {
  "name": "",
  "signature": "sha256:50fdec3117846e114a2ab38e86c74fa28a1d745e07915abb6821cc6f9bf070ae"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "This notebook updates all file annotations for the PCBC project. There are four required file annotations for any assay file uploaded to Synapse:\n",
      "\n",
      "* `UID` - a universal ID for this file\n",
      "* `dataType` - `mRNA`, `miRNA`, or `methylation`\n",
      "* `fileType`: `bam`, `fastq`, etc.\n",
      "* `fileSubType`: is `fileType`-specific; e.g., `bam` can have `mapped` or `unmapped`.\n",
      "\n",
      "This script uses the `dataType` field to get assay-specific files (using a query lookup dictionary `assayToQuery`), and then the `UID` to merge existing file annotations and the table metadata (which is obtained through a lookup dictionarly `assayToMetadataTable`."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import sys\n",
      "import pandas as pd\n",
      "import numpy\n",
      "import logging\n",
      "\n",
      "logger = logging.getLogger('pcbc update annots')\n",
      "logger.setLevel(logging.DEBUG)\n",
      "\n",
      "from multiprocessing.dummy import Pool\n",
      "mp = Pool(10)\n",
      "\n",
      "import synapseHelpers\n",
      "import synapseclient\n",
      "\n",
      "syn = synapseclient.Synapse(skip_checks=True)\n",
      "syn.login(silent=True)\n",
      "\n",
      "dataTypes = ['mRNA', 'miRNA', 'methylation']\n",
      "\n",
      "dataTypesToMetadataTable = dict(mRNA='syn3156503',\n",
      "                                miRNA='syn3219876',\n",
      "                                methylation='syn3156828')\n",
      "\n",
      "dataTypesToQuery = dict(mRNA=\"select id,UID from file where projectId=='syn1773109' AND dataType=='mRNA'\",\n",
      "                        miRNA=\"select id,UID from file where projectId=='syn1773109' AND dataType=='miRNA'\",\n",
      "                        methylation=\"select id,UID from file where projectId=='syn1773109' AND dataType=='methylation' AND fileType=='idat'\")"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def fixDict(d):\n",
      "    \"\"\"Fix dictionaries having .\n",
      "    \n",
      "    \"\"\"\n",
      "    \n",
      "    newd = {}\n",
      "    for k, v in d.iteritems():\n",
      "        tmp = {}\n",
      "        for k2, v2 in v.iteritems():\n",
      "            if v2 is list:\n",
      "                tmp[k2] = v2\n",
      "            else:\n",
      "                try:\n",
      "                    if not numpy.isnan(v2):\n",
      "                        tmp[k2] = v2\n",
      "                except TypeError:\n",
      "                    tmp[k2] = v2\n",
      "        newd[k] = tmp\n",
      "    return newd\n",
      "\n",
      "def updateDict(a, b):\n",
      "    \"\"\"Remove synapse fields from Synapse query on file annotations.\n",
      "    \n",
      "    \"\"\"\n",
      "    \n",
      "    foo = [a.pop(x) for x in [\"creationDate\", \"etag\", \"id\", \"uri\"]]\n",
      "    a.update(b)\n",
      "    return a\n",
      "\n",
      "def updateAnnots(synid, lookupDict):\n",
      "    \"\"\"Update annotations using a dict to lookup which annotations to add (keys are synids).\n",
      "    \n",
      "    Works this way for parallelization.\n",
      "    \n",
      "    \"\"\"\n",
      "    logger.info(\"%s\" % synid)\n",
      "    o = syn.get(synid, downloadFile=False)\n",
      "    a = syn.getAnnotations(o)\n",
      "    newa = a.copy()\n",
      "    newa = updateDict(newa, lookupDict[o.properties.id])\n",
      "    foo = syn.setAnnotations(o, newa)\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 4
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "doupdate = True\n",
      "# updateDatTypes = dataTypes\n",
      "updateDataTypes = ['mRNA']\n",
      "\n",
      "for dataType in updateDataTypes:\n",
      "        \n",
      "    # Metadata\n",
      "    logger.info(\"Getting %s metadata\" % dataType)\n",
      "    metaSchema = syn.get(dataTypesToMetadataTable[dataType])\n",
      "    metaResults = syn.tableQuery(\"select * from %s\" % metaSchema.id)\n",
      "    meta = metaResults.asDataFrame()\n",
      "    \n",
      "    # All files\n",
      "    logger.info(\"Getting %s file list\" % dataType)\n",
      "    fileTbl = synapseHelpers.query2df(syn.chunkedQuery(dataTypesToQuery[dataType]), \n",
      "                                      filterSynapseFields=False)\n",
      "    \n",
      "    # Merge metadata and files\n",
      "    logger.info(\"Merging %s\" % dataType)\n",
      "    \n",
      "    merged = pd.merge(left=fileTbl[['id', 'UID']],  right=meta,\n",
      "                      how=\"left\", left_on=\"UID\", right_on=\"UID\")\n",
      "    \n",
      "    merged.index = merged.id\n",
      "    merged.drop(\"id\", axis=1, inplace=True)\n",
      "    mergedDict = merged.transpose().to_dict()\n",
      "    mergedDict2 = fixDict(mergedDict)\n",
      "    \n",
      "    # Update annotations\n",
      "    if doupdate:\n",
      "        logger.info(\"Updating %s annotations\" % dataType)\n",
      "        res = mp.map(lambda x: updateAnnots(x, mergedDict2), mergedDict2.keys())"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}